### ISSUE TITLE
Observer not unmounted? 

### ISSUE THREAD

**Issue by MartinBspheroid** [wellyshen/react-cool-inview#650]
Hi there,
I cannot figure out why useEffect is trying to [set state here ](https://github.com/wellyshen/react-cool-inview/blob/0e454c17fa7d5714a6deeee46a067e1dd6815954/src/index.ts#L160) after component is unmounted. Even wierder is the fact that this error fires only once. 

![image](https://user-images.githubusercontent.com/3605151/133092473-4c577beb-d2e5-4414-8c0e-bdd0c1436671.png)

Do you have any clue why this would be a case? 

Cheers for the coolest hooks, btw 👍

**Comment by wellyshen** [comment]
@MartinBspheroid Hmm... 🤔, can you share the reproduced code via [CodeSandbox](https://codesandbox.io/)?

**Comment by MartinBspheroid** [comment]
That will require some time, it's part of big codebase. I'll try to make simpified version tommorow and put in up on CodeSanbox. 
Maybe that will reveal source of the problem.

**Comment by wellyshen** [comment]
@MartinBspheroid Thank you mate, that will be helpful 👍🏻

**Comment by MartinBspheroid** [comment]
Alright, so....
[Here's replica of what is pretty much going on](https://codesandbox.io/s/exciting-chatelet-4ifps?file=/src/Cooler.tsx) on CodeSandbox. Funny thing is that within CodeSandbox, it's working alright, but on my testing playground (barebones react/ts/less boilerplate), it's still not happy. 
It's not really crashing react app or not misbehaving in some bad way, it's working fine, but I'd like to keep my console error free :D
Here's a console error from my repro:
![image](https://user-images.githubusercontent.com/3605151/133211036-095293dd-1c1a-4a49-8500-d54d30528030.png)

**Comment by wellyshen** [comment]
@MartinBspheroid I tried to reproduce that in local but I didn't see any error. Did I miss something?

https://user-images.githubusercontent.com/21308003/133237129-7671b834-4d39-45a6-b6a1-2ae64b6bad59.mov

**Comment by MartinBspheroid** [comment]
Okay, you are absolutely right, unfortunately. I 've just grabbed ('off the shelve', so to say) new boiler-plate from github, plugged this repro in and... nothing. Works exactly your. I've even disabled React dev tools, made sure that React versions are the same, even opened annonymous window to compare one I have in different playground, but I cannot reproduce this. I'll give it a shot after lunch again.

**Comment by wellyshen** [comment]
@MartinBspheroid I close this issue atm. Feel free to ask for re-opening once you can reproduce this issue.

**Comment by MartinBspheroid** [comment]
Hey Welly, I was about to close this yesterday on my way from office, but you beat me to it :) Sure, atm I cannot think of any reason why this keeps coming up in the console, but hey hoo, it's still working as expected, so I'm just going to run with it as it is now. 

Cheers for help!

**Comment by h0wXD** [comment]
![image](https://user-images.githubusercontent.com/10574115/141721103-6dedc22a-92ff-4d3b-897d-8e083047c317.png)

I also have a duplicate re-render and whenever the component unmounts, first I see it hit the useEffect cleanup function 
```return () => unobserve();``` after which it'll re-invoke the `IntersectionObserver` and call setState, triggering the error.
Screenshot showing the invocation after unmounting:
![image](https://user-images.githubusercontent.com/10574115/141722199-e597677d-319c-4253-abbf-6fc45fd5323d.png)

**Comment by wellyshen** [comment]
@phenry20 Could you mind to share a minimal reproduce code with me via CodeSandbox?

**Comment by h0wXD** [comment]
@wellyshen Sandbox link https://codesandbox.io/s/dreamy-mcnulty-3jorm?file=/src/App.js
Can simulate by either clicking the title or scrolling.
Although not exactly the same behavior as I experienced, it also does a setState call after hitting the useEffect cleanup function

**Comment by wellyshen** [comment]
@h0wXD Thank you will investigate this issue, soon.

**Comment by wellyshen** [comment]
@h0wXD I copied the code of the [CodeSandbox](https://codesandbox.io/s/dreamy-mcnulty-3jorm?file=/src/App.js) and tried to reproduce this issue on my local environment but I can't reproduce this issue. That's wired, did I miss something?

- My react version: v17.0.2
- Browser version: Chrome 95.0.4638.69

https://user-images.githubusercontent.com/21308003/143381526-e45ce4bf-5577-4b4f-aea4-f3a2e289e5ee.mov

**Comment by robert-w-gries** [comment]
@h0wXD your `loadMore()` method is unmounting `LastItem` with text `Learn JavaScript 25` before the [intersection observer callback](https://github.com/wellyshen/react-cool-inview/blob/master/src/index.ts#L113-L162) finishes.

You can verify this by adding a breakpoint at line 52 of the codesandbox App.js file. Before that line is executed, the `LastItem` component is still mounted. After executing that line, you change the items in the list and cause a re-render. After re-render, the mounted LastItem has text `Learn Javascript 50`, but the observer callback is still executing at this line: https://github.com/wellyshen/react-cool-inview/blob/master/src/index.ts#L152

**Comment by MartinBspheroid** [comment]
So in the end, I've managed to fix this. Turns out that it has nothing to do with intersection observer at all. I've even went with pure vanilla js implementation (it's really just couple of lines for what I really needed), but problem remained. 

Thing that kept triggering that error was async loading function (triggered by observer) that tried to change state of component that was already unmounted. So, applying [this pattern ](https://juliangaramendy.dev/blog/use-promise-subscription) fixed that issue and I'm finally left with no errors in console. Sweet :)
